aws/resource_aws_wafv2_rule_group.go

```diff
@@ -0,0 +1,1361 @@
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"arn": {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validateArn,
+				},
+				"field_to_match":      wafv2FieldToMatchSchema(),
+				"text_transformation": wafv2TextTransformationSchema(),
+			},
+		},
+	}
+}
+
+func wafv2SizeConstraintSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"comparison_operator": {
+					Type:     schema.TypeString,
+					Required: true,
+					ValidateFunc: validation.StringInSlice([]string{
+						wafv2.ComparisonOperatorEq,
+						wafv2.ComparisonOperatorGe,
+						wafv2.ComparisonOperatorGt,
+						wafv2.ComparisonOperatorLe,
+						wafv2.ComparisonOperatorLt,
+						wafv2.ComparisonOperatorNe,
+					}, false),
+				},
+				"field_to_match": wafv2FieldToMatchSchema(),
+				"size": {
+					Type:         schema.TypeInt,
+					Required:     true,
+					ValidateFunc: validation.IntBetween(0, 21474836480),
+				},
+				"text_transformation": wafv2TextTransformationSchema(),
+			},
+		},
+	}
+}
+
+func wafv2SqliMatchStatementSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"field_to_match":      wafv2FieldToMatchSchema(),
+				"text_transformation": wafv2TextTransformationSchema(),
+			},
+		},
+	}
+}
+
+func wafv2XssMatchStatementSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"field_to_match":      wafv2FieldToMatchSchema(),
+				"text_transformation": wafv2TextTransformationSchema(),
+			},
+		},
+	}
+}
+
+func wafv2FieldToMatchSchema() *schema.Schema {
+	return &schema.Schema{
+		Type:     schema.TypeList,
+		Optional: true,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"all_query_arguments": wafv2EmptySchema(),
+				"body":                wafv2EmptySchema(),
+				"method":              wafv2EmptySchema(),
+				"query_string":        wafv2EmptySchema(),
+				"single_header": {
+					Type:     schema.TypeList,
+					Optional: true,
+					MaxItems: 1,
+					Elem: &schema.Resource{
+						Schema: map[string]*schema.Schema{
+							"name": {
+								Type:     schema.TypeString,
+								Required: true,
+								ValidateFunc: validation.All(
+									validation.StringLenBetween(1, 40),
+									// The value is returned in lower case by the API.
+									// Trying to solve it with StateFunc and/or DiffSuppressFunc resulted in hash problem of the rule field or didn't work.
+									validateWafv2StringIsLowerCase(),
```

Comment:
what type of `DiffSuppressFunc` did you experiment with @pvanbuijtene ? I was imagining something like this might do the trick:
```
DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
				return strings.EqualFold(old, new)
},
```

Reply:
@anGie44  I tried it with your suggestion, but the test still results in: `After applying this step and refreshing, the plan was not empty`, I'll attach the output. I tried to get it to work, but couldn't and my suspicion is that it's related to the calculated hash index with the lower and upper case.


I would be happy with any suggestion to get this to work without the lower-case validation :)

Reply:
I can't attach a file for some reason, this is the test output:
```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 1 -run=TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch -timeout 120m
=== RUN   TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch
=== PAUSE TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch
=== CONT  TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch
--- FAIL: TestAccAwsWafv2RuleGroup_ByteMatchStatement_FieldToMatch (42.55s)
    testing.go:683: Step 0 error: After applying this step and refreshing, the plan was not empty:

        DIFF:

        DESTROY/CREATE: aws_wafv2_rule_group.test
          arn:                                                                                "arn:aws:wafv2:eu-central-1:342992114854:regional/rulegroup/rule-group-nitea/5f7ebcac-bb9d-49b8-88b9-73ff2210f243" => "<computed>"
          capacity:                                                                           "15" => "15"
          description:                                                                        "" => ""
          id:                                                                                 "5f7ebcac-bb9d-49b8-88b9-73ff2210f243" => "<computed>"
          lock_token:                                                                         "a49fb90e-02c3-4056-8866-529b6473bf45" => "<computed>"
          name:                                                                               "rule-group-nitea" => "rule-group-nitea"
          rule.#:                                                                             "1" => "1" (forces new resource)
          rule.0.action.#:                                                                    "1" => "1" (forces new resource)
          rule.0.action.0.allow.#:                                                            "1" => "1" (forces new resource)
          rule.0.action.0.block.#:                                                            "0" => "0" (forces new resource)
          rule.0.action.0.count.#:                                                            "0" => "0" (forces new resource)
          rule.0.name:                                                                        "rule-1" => "rule-1" (forces new resource)
          rule.0.priority:                                                                    "1" => "1" (forces new resource)
          rule.0.statement.#:                                                                 "1" => "1" (forces new resource)
          rule.0.statement.0.and_statement.#:                                                 "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.#:                                          "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.#:                         "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.all_query_arguments.#:   "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.body.#:                  "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.method.#:                "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.query_string.#:          "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_header.#:         "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_header.0.name:    "a-forty-character-long-header-name-40-40" => "A-forty-character-long-header-name-40-40" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_query_argument.#: "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.uri_path.#:              "0" => "0" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.positional_constraint:                    "CONTAINS" => "CONTAINS" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.search_string:                            "word" => "word" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.text_transformation.#:                    "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.text_transformation.0.priority:           "1" => "1" (forces new resource)
          rule.0.statement.0.byte_match_statement.0.text_transformation.0.type:               "NONE" => "NONE" (forces new resource)
          rule.0.statement.0.geo_match_statement.#:                                           "0" => "0" (forces new resource)
          rule.0.statement.0.ip_set_reference_statement.#:                                    "0" => "0" (forces new resource)
          rule.0.statement.0.not_statement.#:                                                 "0" => "0" (forces new resource)
          rule.0.statement.0.or_statement.#:                                                  "0" => "0" (forces new resource)
          rule.0.statement.0.regex_pattern_set_reference_statement.#:                         "0" => "0" (forces new resource)
          rule.0.statement.0.size_constraint_statement.#:                                     "0" => "0" (forces new resource)
          rule.0.statement.0.sqli_match_statement.#:                                          "0" => "0" (forces new resource)
          rule.0.statement.0.xss_match_statement.#:                                           "0" => "0" (forces new resource)
          rule.0.visibility_config.#:                                                         "1" => "1" (forces new resource)
          rule.0.visibility_config.0.cloudwatch_metrics_enabled:                              "false" => "false" (forces new resource)
          rule.0.visibility_config.0.metric_name:                                             "friendly-rule-metric-name" => "friendly-rule-metric-name" (forces new resource)
          rule.0.visibility_config.0.sampled_requests_enabled:                                "false" => "false" (forces new resource)
          scope:                                                                              "REGIONAL" => "REGIONAL"
          visibility_config.#:                                                                "1" => "1"
          visibility_config.0.cloudwatch_metrics_enabled:                                     "false" => "false"
          visibility_config.0.metric_name:                                                    "friendly-metric-name" => "friendly-metric-name"
          visibility_config.0.sampled_requests_enabled:                                       "false" => "false"



        STATE:

        aws_wafv2_rule_group.test:
          ID = 5f7ebcac-bb9d-49b8-88b9-73ff2210f243
          provider = provider.aws
          arn = arn:aws:wafv2:eu-central-1:342992114854:regional/rulegroup/rule-group-nitea/5f7ebcac-bb9d-49b8-88b9-73ff2210f243
          capacity = 15
          description =
          lock_token = a49fb90e-02c3-4056-8866-529b6473bf45
          name = rule-group-nitea
          rule.# = 1
          rule.0.action.# = 1
          rule.0.action.0.allow.# = 1
          rule.0.name = rule-1
          rule.0.priority = 1
          rule.0.statement.# = 1
          rule.0.statement.0.byte_match_statement.# = 1
          rule.0.statement.0.byte_match_statement.0.field_to_match.# = 1
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_header.# = 1
          rule.0.statement.0.byte_match_statement.0.field_to_match.0.single_header.0.name = a-forty-character-long-header-name-40-40
          rule.0.statement.0.byte_match_statement.0.positional_constraint = CONTAINS
          rule.0.statement.0.byte_match_statement.0.search_string = word
          rule.0.statement.0.byte_match_statement.0.text_transformation.# = 1
          rule.0.statement.0.byte_match_statement.0.text_transformation.0.priority = 1
          rule.0.statement.0.byte_match_statement.0.text_transformation.0.type = NONE
          rule.0.visibility_config.# = 1
          rule.0.visibility_config.0.cloudwatch_metrics_enabled = false
          rule.0.visibility_config.0.metric_name = friendly-rule-metric-name
          rule.0.visibility_config.0.sampled_requests_enabled = false
          scope = REGIONAL
          visibility_config.# = 1
          visibility_config.0.cloudwatch_metrics_enabled = false
          visibility_config.0.metric_name = friendly-metric-name
          visibility_config.0.sampled_requests_enabled = false
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       43.060s
FAIL
make: *** [GNUmakefile:26: testacc] Error 1

```

Reply:
ah i see, no worries thanks for the output! going to try on my end and see if maybe it's because the value is quite nested in that statement struct

Reply:
I see what you were saying regarding the hash index 😢 perhaps this `StateFunc` schema attribute could be the alternative https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#statefunc we seek

Reply:
>  but a hash function
I was trying to prevent this, as it is quite some cumbersome work 😬 

Is this not a bug (missing feature) in the terraform code ?

Reply:
Using the `StateFunc` results in other "weird" behavior `panic: missing field in set: rule.486868203.statement.0.byte_match_statement.0.field_to_match.0.single_header.0.name.486868203` it can't find the field anymore 🤔 

I think the current schema code is not entirely prepared for a schema like this with it's API requirements/behavior.

I'm not sure if using a hash func on the rule level is the proper solution for some problem with a field this deep in the schema, can @bflad help with some guidance here 😬

Reply:
hi @pvanbuijtene, asking about this internally, it's a valid work-around to use the `validationfunc` in this case as suggested by the schema's nested nature and statefuncs will require adjustments further up the schema. with that being said, let's use the validation func's in place but if you don't mind can we adjust the message to include `must contain only lowercase...` ?